Add support for specifying metadata options in MDX comments

bunchesofdonald opened the pull request:
### Summary

This PR adds support for specifying test metadata in `.mdx` files using MDX-style comments (`{/* pmd-metadata: */}`) placed above code blocks. This should support all existing options and allow for easily adding more without having to alter the MDX parsing logic.

### Example:
```mdx
{/* pmd-metadata: notest, fixture:capsys */}
```python
print("hello")
captured = capsys.readouterr()
assert captured.out == "hello\n"
```

* Updated documentation with MDX metadata usage.
* Minor refactoring for readability and metadata extraction.

**Issue:** https://github.com/modal-labs/pytest-markdown-docs/issues/28

bunchesofdonald commented:
@freider or @mwaskom could I get a review of this? I'd really love to use this module, but this is a big blocker for that.

freider commented on src/pytest_markdown_docs/plugin.py:
```diff
@@ -159,29 +162,55 @@ def extract_code_blocks(
+        # MDX comments are put inside of a paragraph block, so we check the
+        # block two back to see if it's a comment as we the comment needs to
+        # be directly above the code block.
+        if i >= 2 and is_mdx_comment(tokens[i - 2]):
```
Would it be messy to make this apply only for `.mdx` contexts and not regular `.md` markdown? In regular markdown the syntax isn't considered an invisible comment afaik, so it shouldn't be treated as metadata.
It would probably require us to propagate that metadata through `MarkdownTextFile` etc.

bunchesofdonald replied:
I think I can probably make that happen.

freider commented on src/pytest_markdown_docs/plugin.py:
```diff
@@ -159,29 +162,55 @@ def extract_code_blocks(
         lang = code_info[0] if code_info else None
+        code_options = set(code_info) - {lang}
+
+        # MDX comments are put inside of a paragraph block, so we check the
```
Could you clarify a bit in the comment why it's two blocks back? What's the order of blocks here?

N - 2: paragraph (comment)
N - 1: ???
N: code fence

bunchesofdonald replied:
It's been a bit since I wrote this, but as I remember it was something like:

```
<paragraph>
 <mdx comment>
</paragraph>
<code fence>
```

So it's `code fence` - 2.

I'll see if I can make the comment more clear.

freider commented on tests/plugin_test.py:
```diff
@@ -223,3 +223,97 @@ def foo():
+
+def test_continuation_mdx_comment(testdir):
+    testdir.makefile(
+        ".md",
```
This should be `.mdx` right?

bunchesofdonald replied:
Yep, thanks.

freider reviewed (commented):
Hi! Thanks for contributing! Feels related to https://github.com/modal-labs/pytest-markdown-docs/pull/26 (a bit annoying that markdown variants haven't settled on a single metadata/comment notation)

As long as it can't accidentally break regular `.md` specifications I'd be happy to add it. Might want to refactor it slightly to make the parser-specific metadata injection more configurable (or dependent on the .mdx context in this case)

freider commented:
An alternative to having the `mdx` context propagate in would be to have another pytest marker/arg "--markdown-docs-syntax=mdx-comments" (I just suggested something similar for the superfences syntax that #26 is adding)

bunchesofdonald commented:
> An alternative to having the `mdx` context propagate in would be to have another pytest marker/arg "--markdown-docs-syntax=mdx-comments" (I just suggested something similar for the superfences syntax that #26 is adding)

Do you have a preference? I think having it pick up this style from the fact it's parsing an `.mdx` file makes sense, but if you feel strongly that having an arg is the way to go that also works for me. Since I think the context would be just a bit friendlier I'll start on that version, but should be easy-ish to change.

freider commented:
> > An alternative to having the `mdx` context propagate in would be to have another pytest marker/arg "--markdown-docs-syntax=mdx-comments" (I just suggested something similar for the superfences syntax that #26 is adding)
> 
> Do you have a preference? I think having it pick up this style from the fact it's parsing an `.mdx` file makes sense, but if you feel strongly that having an arg is the way to go that also works for me.
> 
> Since I think the context would be just a bit friendlier I'll start on that version, but should be easy-ish to change.

The argument for an arg rather than context is that it's more Mintlify-specific than `.mdx`-specific, but I can see the point for having it as an always-on-thing for .mdx files anyways, just in case there are other renderers than Mintlify that use code fence info blocks for rendered content. Also could be quite useful to have file extension context available in the block extractor anyways, so go for it :)

bunchesofdonald commented:
@freider I've addressed your comments, please have a look at the updated code when you have a moment. Let me know if there are any further adjustments or improvements you'd like me to make.

Thanks again for your guidance!

freider commented:
@bunchesofdonald release v0.6.0 includes this patch and is up on pypi now FYI

bunchesofdonald commented:
Amazing, thank you so much @freider!
